src/main/java/com/cvsgo/repository/ProductCustomRepositoryImpl.java

```diff
@@ -0,0 +1,112 @@
+import com.cvsgo.dto.product.QProductResponseDto;
+import com.cvsgo.entity.Category;
+import com.cvsgo.entity.ConvenienceStore;
+import com.cvsgo.entity.EventType;
+import com.cvsgo.entity.QEvent;
+import com.cvsgo.entity.User;
+import com.querydsl.core.BooleanBuilder;
+import com.querydsl.core.types.ExpressionUtils;
+import com.querydsl.core.types.dsl.BooleanExpression;
+import com.querydsl.jpa.JPAExpressions;
+import com.querydsl.jpa.impl.JPAQueryFactory;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+@RequiredArgsConstructor
+@Repository
+public class ProductCustomRepositoryImpl implements ProductCustomRepository {
+
+    private final JPAQueryFactory queryFactory;
+
+    public Page<ProductResponseDto> searchByFilter(User user, ProductSearchFilter filter,
+        Pageable pageable) {
+        List<ProductResponseDto> results = queryFactory
+            .select(new QProductResponseDto(
+                product.id.as("productId"),
+                product.name.as("productName"),
+                product.price.as("productPrice"),
+                product.imageUrl.as("productImageUrl"),
+                product.category.id.as("categoryId"),
+                product.manufacturer.name.as("manufacturerName"),
+                ExpressionUtils.as(JPAExpressions.selectOne().from(productLike)
+                    .where(product.id.eq(productLike.product.id)
+                        .and(productLike.user.id.eq(user.getId()))).limit(1), "isLike"),
+                ExpressionUtils.as(JPAExpressions.selectOne().from(productBookmark)
+                    .where(product.id.eq(productBookmark.product.id)
+                        .and(productBookmark.user.id.eq(user.getId()))).limit(1), "isBookmark"),
+                ExpressionUtils.as(JPAExpressions.select(review.count()).from(review)
+                    .where(product.id.eq(review.product.id)), "reviewCount"),
+                ExpressionUtils.as(JPAExpressions.select(review.rating.avg()).from(review)
+                    .where(product.id.eq(review.product.id)), "reviewRating")
+            ))
+            .from(product)
+            .leftJoin(sellAt).on(sellAt.product.id.eq(product.id))
+            .leftJoin(event).on(event.product.id.eq(product.id))
+            .where(
+                eqConvenienceStore(filter.getConvenienceStore()),
+                eqCategory(filter.getCategory()),
+                eqEvent(filter.getEvent()),
+                priceLesserEqual(filter.getPrice()),
+                priceGreaterOrEqual(filter.getPrice())
+            )
+            .groupBy(product.id)
+            .offset(pageable.getOffset())
+            .limit(pageable.getPageSize())
+            .fetch();
+
+        return new PageImpl<>(results, pageable, results.size());
```

Comment:
PageImpl 생성자의 세 번째 매개변수는 total로, results.size() 대신 총 데이터의 개수가 들어가야 합니다.
원래는 fetch() 대신 fetchResults()를 사용하고 results.getTotal() 하면 되었었는데, QueryDsl 5.0 이상(Spring Boot 2.6 이상)부터 fetchResults() 메소드가 deprecated 되었다고 합니다.
따라서 아래와 같이 별도의 count 쿼리를 통해 총 개수를 구하시면 될 것 같습니다.
```suggestion
        Long totalCount= queryFactory.select(product.count())
                .from(product)
                .fetchOne();
        return new PageImpl<>(results, pageable, totalCount);
```

Reply:
리뷰 감사합니다 이 부분에 대해 더 찾아보도록 하겠습니다.